core/primitives/src/epoch_manager.rs

```diff
@@ -620,42 +618,158 @@ pub struct ValidatorWeight(ValidatorId, u64);
         checked_feature,
         hash::hash,
         types::{BlockHeight, ShardId},
     };
 
-    pub use super::EpochInfoV1;
-
     /// Information per epoch.
+    /// DEPRECATED. Remove after DB version 40 and protocol feature
+    /// StatelessValidationV0 are released.
+    #[derive(
+        SmartDefault,
+        BorshSerialize,
+        BorshDeserialize,
+        Clone,
+        Debug,
+        PartialEq,
+        Eq,
+        serde::Serialize,
+    )]
+    pub struct EpochInfoV1 {
+        /// Ordinal of given epoch from genesis.
+        /// There can be multiple epochs with the same ordinal in case of long forks.
+        pub epoch_height: EpochHeight,
+        /// List of current validators.
+        pub validators: Vec<ValidatorStakeV1>,
+        /// Validator account id to index in proposals.
+        pub validator_to_index: HashMap<AccountId, ValidatorId>,
+        /// Settlement of validators responsible for block production.
+        pub block_producers_settlement: Vec<ValidatorId>,
+        /// Per each shard, settlement validators that are responsible.
+        pub chunk_producers_settlement: Vec<Vec<ValidatorId>>,
+        /// Settlement of hidden validators with weights used to determine how many shards they will validate.
+        pub hidden_validators_settlement: Vec<ValidatorWeight>,
+        /// List of current fishermen.
+        pub fishermen: Vec<ValidatorStakeV1>,
+        /// Fisherman account id to index of proposal.
+        pub fishermen_to_index: HashMap<AccountId, ValidatorId>,
+        /// New stake for validators.
+        pub stake_change: BTreeMap<AccountId, Balance>,
+        /// Validator reward for the epoch.
+        pub validator_reward: HashMap<AccountId, Balance>,
+        /// Validators who are kicked out in this epoch.
+        pub validator_kickout: HashMap<AccountId, LegacyValidatorKickoutReasonV39>,
+        /// Total minted tokens in the epoch.
+        pub minted_amount: Balance,
+        /// Seat price of this epoch.
+        pub seat_price: Balance,
+        /// Current protocol version during this epoch.
+        pub protocol_version: ProtocolVersion,
+    }
+
     #[derive(BorshSerialize, BorshDeserialize, Clone, Debug, PartialEq, Eq, serde::Serialize)]
-    pub enum EpochInfo {
+    /// DEPRECATED. Remove after DB version 40 and protocol feature
+    /// StatelessValidationV0 are released.
+    pub enum LegacyEpochInfo {
         V1(EpochInfoV1),
         V2(EpochInfoV2),
         V3(EpochInfoV3),
         V4(EpochInfoV4),
     }
 
+    /// Information per epoch.
+    /// TODO: V1-V3 are needed for consistent serialization with statelessnet.
+    /// Remove and migrate again once statelessnet is torn down.
+    #[derive(BorshSerialize, BorshDeserialize, Clone, Debug, PartialEq, Eq, serde::Serialize)]
+    pub enum EpochInfo {
+        #[allow(unused)]
+        V1,
+        #[allow(unused)]
+        V2,
+        #[allow(unused)]
+        V3,
+        V4(EpochInfoV4),
```

Comment:
Wait, if we are migrating EpochInfo with DB migration, do we even need to maintain this struct with V1, V2, V3, etc.?

Could the new EpochInfo (after migration) just be a normal serialized struct instead of being versioned?

Reply:
We don't need to maintain them, but I want this layout to be compatible with statelessnet. I don't want to migrate statelessnet and it uses versioned enum.
After statelessnet is removed, we can add another migration and remove V1-V3.